merkle/src/merkle.rs

```diff
@@ -586,72 +623,216 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>> MerkleTree<T, A, K> {
+                (leaves_lock.clone(), top_half_lock.clone(), 0, 0)
+            } else {
+                // For all other levels we'll read/write from/to the `top_half` adjusting the
+                // "global" index to access this `Store` (offsetting `leaves` length). All levels
+                // are contiguous so we read/write `width` nodes apart.
+                let read_start = level_node_index - leaves_lock.read().unwrap().len();
+                (
+                    top_half_lock.clone(),
+                    top_half_lock.clone(),
+                    read_start,
+                    read_start + width,
+                )
+            };
+            // FIXME: Maybe just remove `write_store_lock` and always access `top_half_lock`
+            // directly if it makes it more readable.
+
+            // Allocate `width` indexes during operation (which is a negligible memory bloat
+            // compared to the 32-bytes size of the nodes stored in the `Store`s) and hash each
+            // pair of nodes to write them to the next level in concurrent threads.
+            // Process `chunk_size` nodes in each thread at a time to reduce contention, optimized
+            // for big sector sizes (small ones will just have one thread doing all the work).
+            let chunk_size = 1024;
+            debug_assert_eq!(chunk_size % 2, 0);
+            Vec::from_iter((read_start..read_start + width).step_by(chunk_size))
+                .par_chunks(1)
+                .for_each(|chunk_index| {
+                    let chunk_index = chunk_index[0];
+                    let chunk_size = std::cmp::min(chunk_size, read_start + width - chunk_index);
+
+                    let chunk_nodes = {
+                        // Read everything taking the lock once.
+                        let read_store = read_store_lock.read().unwrap();
+                        read_store.read_range(std::ops::Range {
+                            start: chunk_index,
+                            end: chunk_index + chunk_size,
+                        })
+                    };
+
+                    let hashed_nodes = chunk_nodes.chunks(2).map(|node_pair| {
+                        A::default().node(node_pair[0].clone(), node_pair[1].clone(), level)
+                        // FIXME: Change `node()` to receive references to avoid the `clone` here,
+                        //  this might be an API-breaking change.
+                    });
+
+                    // We write the hashed nodes to the next level in the position that
+                    // would be "in the middle" of the previous pair (dividing by 2).
+                    let write_delta = (chunk_index - read_start) / 2;
+
+                    let mut hashed_nodes_as_bytes: Vec<u8> =
+                        Vec::with_capacity(hashed_nodes.len() * T::byte_len());
+                    hashed_nodes
+                        .for_each(|node| hashed_nodes_as_bytes.append(&mut node.as_ref().to_vec()));
```

Comment:
the above can be written a bit nicer, to allow rust to optimize the code potentially better I believe

```rust
let nodes_size = (chunk_nodes.len() / 2) * T::byte_len();
let hashed_nodes_as_bytes = chunk_nodes.chunks(2).fold(
    Vec::with_capacity(nodes_size), 
    |mut acc, node_pair| {
		let h = A::default().node(node_pair[0].clone(), node_pair[1].clone(), level);
        acc.extend_from_slice(h.as_ref());
		acc
    }
);
```

Reply:
it should also avoid the one allocation per hash (of the `to_vec()`)